### PR TITLE
implement redis

### PR DESCRIPTION
**Pull request by Dhruvi-Makwana**

### CODE REVIEW

File: Chatify/Chatify/settings/base.py [hunk 1]
```diff
@@ -7,6 +7,9 @@
 import os
 import environ
 
+from celery.schedules import crontab
```
Comment (yatishTT) [inline Chatify/Chatify/settings/base.py]: Why do we need this

File: Chatify/chat/api.py [hunk 2]
```diff
@@ -1,7 +1,7 @@
 from rest_framework.views import APIView
 from rest_framework.response import Response
 from rest_framework import status
-from .serializers import UserSerializer, LoginSerializer
+from .serializers import UserSerializer, LoginSerializer, UpdateUserActiveTime
```
Comment (yatishTT) [inline Chatify/chat/api.py]: UpdateUserLastActiveTime

File: Chatify/chat/redis.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+def check_last_login(user_id):
+
+    current_user = User.objects.get(id=user_id)
+    utc_now = timezone.now()
+    time_diff = utc_now - current_user.last_login
+    current_user.last_login = utc_now
+    current_user.save()
+    print(current_user.last_login)
+    if time_diff.total_seconds() > 20:
```
Comment (yatishTT) [inline Chatify/chat/redis.py]: remove this if 

why we need this

File: Chatify/chat/api.py [hunk 4]
```diff
@@ -84,3 +85,10 @@ def get(self, request):
         send_chat_message(set_status(request.user), "logout")
         logout(request)
         return redirect(reverse("chat:loginUI"))
+
+
+class SetUserActiveTime(APIView):
+    def get(self, request):
+        user = User.objects.filter(username=request.user)
```
Comment (yatishTT) [inline Chatify/chat/api.py]: request.user.id

File: Chatify/chat/api.py [hunk 5]
```diff
@@ -84,3 +85,10 @@ def get(self, request):
...
+        check_last_login(request.user.id)
+        return JsonResponse({"user": list(UpdateUserActiveTime(user, many=True).data)})
```
Comment (yatishTT) [inline Chatify/chat/api.py]: it will return 200 code / "ok"

File: Chatify/chat/redis.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+from .websocket_utils import send_chat_message
+from django.utils import timezone
+import redis
+
+
+def check_last_login(user_id):
+
+    current_user = User.objects.get(id=user_id)
+    utc_now = timezone.now()
+    time_diff = utc_now - current_user.last_login
```
Comment (yatishTT) [inline Chatify/chat/redis.py]: it should set in redis

File: Chatify/chat/static/js/chat.js [hunk 7]
```diff
@@ -96,4 +96,13 @@ app.controller('chatCtrl', function($scope, $http) {
             setUserStatus($scope.status, $scope.id)
         })
     }
+
+    $scope.myInterval = setInterval(getUserLastActiveTime, 20000);
+
+    function getUserLastActiveTime()
```
Comment (yatishTT) [inline Chatify/chat/static/js/chat.js]: setUserLastActiveTime

File: Chatify/chat/serializers.py [hunk 8]
```diff
@@ -57,3 +57,12 @@ def validate_password(self, password):
 class LoginSerializer(serializers.Serializer):
     username = serializers.CharField()
     password = serializers.CharField()
+
+
+class UpdateUserActiveTime(serializers.ModelSerializer):
```
Comment (yatishTT) [inline Chatify/chat/serializers.py]: Why we need this ?

**Review by yatishTT — CHANGES_REQUESTED** [review]

File: Chatify/chat/redis.py [hunk 9]
```diff
@@ -0,0 +1,13 @@
+from .websocket_utils import send_chat_message
+import redis
+from django.utils import timezone
```
Comment (yatishTT) [inline Chatify/chat/redis.py]: FILE_NAME = redis_utils.py

File: Chatify/chat/redis.py [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+import pytz
+
+
+def check_last_login(user_id):
+    r = redis.Redis()
+    last_login = r.get(f"user:{user_id}:last_login")
+    last_update_time = timezone.datetime.fromtimestamp(float(last_login), pytz.utc)
```
Comment (yatishTT) [inline Chatify/chat/redis.py]: from django.utils import timezone
last_update_time = timezone.now()

File: Chatify/chat/redis.py [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+    current_dt = timezone.now().astimezone(pytz.utc)
+    send_chat_message(user_id, "login")
```
Comment (yatishTT) [inline Chatify/chat/redis.py]: Don't set this send_chat_message(user_id, "login")

File: Chatify/chat/redis.py [hunk 12]
```diff
@@ -0,0 +1,13 @@
+from .websocket_utils import send_chat_message
+import redis
+from django.utils import timezone
+import pytz
+
+
+def check_last_login(user_id):
+    r = redis.Redis()
+    last_login = r.get(f"user:{user_id}:last_login")
```
Comment (yatishTT) [inline Chatify/chat/redis.py]: user_last_active_time_key_identifier = f"user:{user_id}:last_login"

File: Chatify/chat/redis.py [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+    last_update_time = timezone.datetime.fromtimestamp(float(last_login), pytz.utc)
+    current_dt = timezone.now().astimezone(pytz.utc)
+    send_chat_message(user_id, "login")
+    r.set(f"user:{user_id}:last_login", timezone.now().timestamp())
```
Comment (yatishTT) [inline Chatify/chat/redis.py]: r.set(user_last_active_time_key_identifier, last_update_time)

**Review by yatishTT — CHANGES_REQUESTED** [review]

File: Chatify/chat/redis.py [hunk 14]
```diff
@@ -0,0 +1,13 @@
+from .websocket_utils import send_chat_message
+import redis
+from django.utils import timezone
+import pytz
+
+
```
Comment (yatishTT) [inline Chatify/chat/redis.py]: It is a global variable 


REDIS_CACHE = redis.Redis()

**Review by yatishTT — CHANGES_REQUESTED** [review]

File: Chatify/chat/redis.py (hunk 12 above)
Comment (Dhruvi-Makwana) [reply Chatify/chat/redis.py]: can we write this big user_last_active_time_key_identifier variable name ?

File: Chatify/chat/api.py [hunk 15]
```diff
@@ -84,3 +87,18 @@ def get(self, request):
         send_chat_message(set_status(request.user), "logout")
         logout(request)
         return redirect(reverse("chat:loginUI"))
+
+
+class SetUserActiveTime(APIView):
+    def get(self, request):
+        check_last_login(request.user.id)
+        return JsonResponse(
```
Comment (yatishTT) [inline Chatify/chat/api.py]: return "ok"

File: Chatify/chat/redis_utils.py [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+from django.utils import timezone
+
+REDIS_CACHE = redis.Redis()
+
+
+def check_last_login(user_id):
+    r = redis.Redis()
+    user_last_active_time_key_identifier = f"user:{user_id}:last_login"
+    last_update_time = timezone.now().isoformat()
+    r.set(user_last_active_time_key_identifier, last_update_time)
```
Comment (yatishTT) [inline Chatify/chat/redis_utils.py]: REDIS_CACHE.set

**Review by yatishTT — CHANGES_REQUESTED** [review]

**Review by yatishTT — APPROVED** [review]

**Pull request merged**